esmvaltool/diag_scripts/lifetime/lifetime_base.py

```diff
@@ -0,0 +1,875 @@
+    latitude : iris.cube.Coord
+        The latitude coordinate of the grid.
+
+    longitude : iris.cube.Coord
+        The longitude coordinate of the grid.
+
+    Returns
+    -------
+    iris.cube
+        An array with the area (in m2) and the input latitude and longitude as
+        coordinates.
+    """
+    r_earth = 6378100.0  # from astropy.constants
+
+    lat_i = np.deg2rad(guess_interfaces(latitude.points))
+    lon_i = np.deg2rad(guess_interfaces(longitude.points))
+
+    delta_x = abs(lon_i[1:] - lon_i[:-1])
+    delta_y = abs(np.sin(lat_i[1:]) - np.sin(lat_i[:-1]))
+
+    output = np.outer(delta_y, delta_x) * r_earth**2
+    output = output.astype('float32')
+
+    result = iris.cube.Cube(output,
+                            standard_name='cell_area',
+                            long_name='cell_area',
+                            var_name='cell_area',
+                            units='m2',
+                            dim_coords_and_dims=[(latitude, 0),
+                                                 (longitude, 1)])
+
+    return result
+
+
+def dpres_plevel_4d(plev, pmin, pmax, z_coord='air_pressure'):
+    """Calculate delta pressure levels.
+
+    The delta pressure levels are based
+    on the given pressure level as a
+    four dimensional cube.
+
+    """
+    cubelist_dplev = [plev_slice.copy()
+                      for plev_slice in plev.slices(['time',
+                                                     'latitude',
+                                                     'longitude'],
+                                                    ordered=True)]
+    cubelist_plev = [plev_slice.copy()
+                     for plev_slice in plev.slices(['time',
+                                                    'latitude',
+                                                    'longitude'],
+                                                   ordered=True)]
+
+    increasing = (plev.coord(z_coord,
+                             dim_coords=True).attributes['positive'] == 'down')
+    last = plev.coords(z_coord)[0].shape[0] - 1
+
+    for i, lev in enumerate(cubelist_plev):
+        if increasing:
+            increment = [i + 1, i - 1]
+        else:
+            increment = [i - 1, i + 1]
+
+        if i == 0:
+            cube = (cubelist_plev[increment[0]] - lev) / 2. + (lev - pmin)
+            cubelist_dplev[i].data = cube.data
+        elif i == last:
+            cube = (pmax - lev) + (lev - cubelist_plev[increment[1]]) / 2.
+            cubelist_dplev[i].data = cube.data
+        else:
+            cube = ((lev - cubelist_plev[increment[0]]) / 2.
+                    + (cubelist_plev[increment[1]] - lev) / 2.)
+            cubelist_dplev[i].data = cube.data
+
+        cubelist_dplev[i].add_aux_coord(iris.coords.AuxCoord(
+            lev.coords(z_coord)[0].points[0]))
+
+    dplev = iris.cube.CubeList(cubelist_dplev).merge_cube()
+    dplev.transpose(new_order=[1, 0, 2, 3])
+    dplev = iris.util.reverse(dplev, 1)
```

Comment:
Creating these copies of the cubes and then iterating over them is inefficient and expensive. I think it should be possible to directly operate on the arrays here. We have some code [here](https://github.com/ESMValGroup/ESMValCore/blob/f72a873a4fcd92363bcf785f343dcd8562ac9810/esmvalcore/preprocessor/_derive/_shared.py#L170) that tries to do that, but unfortunately that uses numpy arrays instead of dask. It basically boils down to creating a copy of the array with indices shifted by one (using `dask.array.roll`) and then add this to the original array.